src/nile/core/types/account.py

```diff
@@ -0,0 +1,288 @@
+"""Account module."""
+import logging
+import os
+
+from dotenv import load_dotenv
+
+from nile import accounts, deployments
+from nile.common import (
+    NILE_ARTIFACTS_PATH,
+    UNIVERSAL_DEPLOYER_ADDRESS,
+    is_alias,
+    normalize_number,
+)
+from nile.core.types.transactions import (
+    DeclareTransaction,
+    DeployAccountTransaction,
+    InvokeTransaction,
+)
+from nile.core.types.tx_wrappers import (
+    DeclareTxWrapper,
+    DeployAccountTxWrapper,
+    DeployContractTxWrapper,
+    InvokeTxWrapper,
+)
+from nile.core.types.udc_helpers import create_udc_deploy_transaction
+from nile.core.types.utils import get_counterfactual_address, get_execute_calldata
+from nile.utils.get_nonce import get_nonce_without_log as get_nonce
+
+try:
+    from nile.core.types.signer import Signer
+except ImportError:
+    pass
+
+load_dotenv()
+
+
+class AsyncObject(object):
+    """Base class for Account to allow async initialization."""
+
+    async def __new__(cls, *a, **kw):
+        """Return coroutine (not class so sync __init__ is not invoked)."""
+        instance = super().__new__(cls)
+        await instance.__init__(*a, **kw)
+        return instance
+
+    async def __init__(self):
+        """Support Account async __init__."""
+        pass
+
+
+class Account(AsyncObject):
+    """
+    Account contract abstraction.
+
+    Remove AsyncObject if Account.deploy decouples from initialization.
+    """
+
+    async def __init__(
+        self,
+        signer,
+        network,
+        salt=0,
+        max_fee=None,
+        predeployed_info=None,
+        watch_mode=None,
+        auto_deploy=True,
+    ):
+        """Get or deploy an Account contract for the given private key."""
+        signer, alias = _get_signer_and_alias(signer, predeployed_info)
+
+        self.signer = signer
+        self.alias = alias
+        self.network = network
+
+        if predeployed_info is not None:
+            self.address = predeployed_info["address"]
+            self.index = predeployed_info["index"]
+        elif accounts.exists(self.signer.public_key, network):
+            signer_data = next(accounts.load(self.signer.public_key, network))
+            self.address = signer_data["address"]
+            self.index = signer_data["index"]
+        elif auto_deploy:
+            tx = await self.deploy(salt=salt, max_fee=max_fee)
+            # DeployAccountTxWrapper.execute updates account's address and index
+            await tx.execute(watch_mode=watch_mode)
+
+        # We should replace this with static type checks
+        if hasattr(self, "address"):
+            assert type(self.address) == int
+
+    async def deploy(self, salt=None, max_fee=None, abi=None):
+        """Deploy an Account contract for the given private key."""
+        salt = 0 if salt is None else normalize_number(salt)
+        calldata = [self.signer.public_key]
+        contract_name = "Account"
+        predicted_address = get_counterfactual_address(salt, calldata)
```

Comment:
Removed `predicted_address` from wrapper constructor.

Regarding salt, calldata, and contract_name, I don't think they should be handled by `DeployAccountTransaction`, because this could be used to deploy a different account contract, and then the calldata and the name could be different. I think is make sense to keep this logic specific to our account, in the account.